code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -715,75 +711,121 @@ pub mod pallet {
+
+				// REVIEW(benluelo): Is this check necessary/ worth it?
+				ensure!(
+					matches!(
+						T::Assets::can_withdraw(rewards_pool.asset_id, who, amount),
+						WithdrawConsequence::Success
+					),
+					Error::<T>::NotEnoughAssets
+				);
 
-			ensure!(
-				matches!(
-					T::Assets::can_withdraw(rewards_pool.asset_id, who, amount),
-					WithdrawConsequence::Success
-				),
-				Error::<T>::NotEnoughAssets
-			);
+				// REVIEW(benluelo): What does "boosted amount" mean?
+				// REVIEW(benluelo): Make this a method on `Lock`
+				let boosted_amount = rewards_pool
+					.lock
+					.duration_presets
+					.get(&duration_preset)
+					.cloned()
+					// TODO(benluelo): Rename this error to something like "DurationPresetNotFound"
+					.ok_or(Error::<T>::NoDurationPresetsConfigured)?
+					.mul_ceil(amount);
+
+				let mut reductions = BTreeMap::new();
+
+				for (asset_id, reward) in
+					(&mut rewards_pool.rewards).into_iter().filter_map(|(asset_id, reward)| {
+						match reward {
+							RewardType::Earnings() => None,
+							RewardType::RateBased(ref mut rate_based_reward) =>
+								Some((asset_id, rate_based_reward)),
+						}
+					}) {
+					let inflation = if rewards_pool.total_shares == T::Balance::zero() {
+						T::Balance::zero()
+					} else {
+						reward
+							.total_rewards
+							.safe_mul(&boosted_amount)?
+							.safe_div(&rewards_pool.total_shares)?
+					};
+
+					reward.total_rewards = reward.total_rewards.safe_add(&inflation)?;
+					reward.total_dilution_adjustment =
+						reward.total_dilution_adjustment.safe_add(&inflation)?;
+
+					reductions.insert(*asset_id, inflation);
+				}
 
-			let boosted_amount = Self::boosted_amount(reward_multiplier, amount);
-			let (rewards, reductions) =
-				Self::compute_rewards_and_reductions(boosted_amount, &rewards_pool)?;
-
-			let fnft_collection_id = rewards_pool.financial_nft_asset_id;
-			let fnft_instance_id = T::FinancialNft::get_next_nft_id(&fnft_collection_id)?;
-			let fnft_account =
-				T::FinancialNft::asset_account(&fnft_collection_id, &fnft_instance_id);
-
-			let new_position = StakeOf::<T> {
-				reward_pool_id: *pool_id,
-				stake: amount,
-				share: boosted_amount,
-				reductions,
-				lock: lock::Lock {
-					started_at: T::UnixTime::now().as_secs(),
-					duration: duration_preset,
-					unlock_penalty: rewards_pool.lock.unlock_penalty,
-				},
-				fnft_instance_id,
-			};
+				let fnft_instance_id =
+					T::FinancialNft::get_next_nft_id(&rewards_pool.fnft_collection_id)?;
+				let fnft_account = T::FinancialNft::asset_account(
+					&rewards_pool.fnft_collection_id,
+					&fnft_instance_id,
+				);
 
-			rewards_pool.total_shares = rewards_pool.total_shares.safe_add(&boosted_amount)?;
-			rewards_pool.rewards = rewards;
+				// REVIEW(benluelo): Is this fallible? What does it mean if this fails?
+				rewards_pool.total_shares = rewards_pool.total_shares.safe_add(&boosted_amount)?;
```

Comment:
Only for overflow safety

Reply:
the issue is that if this does overflow then i'm not sure what that means for the pool - using safe addition here prevents it from crashing the node, but the pool might not be in a valid state at that point

Reply:
This does not ignore the overflow, just returns an `ArithmeticError`

Reply:
but is this state reachable? should it be bounded somehow? what would it mean if the total shares hits the max?

Reply:
This could be reached depending on the config and boosting etc. I guess. So it's ok to assume the bound is the maximum allowed here by the type IMO as we are not going to make a logical error anyway.

Reply:
sounds good, i'll document the limit as such